### PR TITLE
[AutoDiff] Fix logic for checking differentiation generic requirements.

### PR DESCRIPTION
**Pull request by dan-zheng**
Use simple, robust logic for checking whether differentiation generic
requirements are satisfied. Substitute requirement types using the given
substitution map, looking up conformances in the current module.

Resolves [TF-508](https://bugs.swift.org/browse/TF-508). Unblocks https://github.com/apple/swift/pull/24786.

### CODE REVIEW

**Comment by dan-zheng** [comment]
@swift-ci Please test tensorflow

**Review by rxwei — APPROVED** [review]
Yay!

File: lib/SILOptimizer/Mandatory/Differentiation.cpp [hunk 1]
```diff
@@ -789,73 +789,44 @@ static bool checkRequirementsSatisfied(
...
-        
-      if (!conformance)
+      // If the first type does not conform to the second type in the current
+      // module, then record the unsatisfied requirement.
+      if (!swiftModule->lookupConformance(firstType, protocol))
         unsatisfiedRequirements.push_back(req);
       continue;
     }
     // Ignore other requirements (superclass and layout).
+    // Layout requirements are rejected during type-checking.
```
Comment (slavapestov) [inline lib/SILOptimizer/Mandatory/Differentiation.cpp]: Is there no value in allowing layout requirements here? The only kind that is implemented so far is conformance to AnyObject, which seems useful.

**Review by slavapestov — COMMENTED** [review]
This is much cleaner, thanks!

File: lib/SILOptimizer/Mandatory/Differentiation.cpp (hunk 1 above)
Comment (dan-zheng) [reply lib/SILOptimizer/Mandatory/Differentiation.cpp]: Is it the case that `T : AnyObject` is a layout requirement, not a conformance requirement?

It's not clear that supporting layout requirements is desirable for differentiation, but we're happy to revisit it after differentiation has better support for classes, and when differentiation is pitched via Swift Evolution.

Comment (slavapestov) [reply lib/SILOptimizer/Mandatory/Differentiation.cpp]: Correct. `T : AnyObject` is a layout requirement. AnyObject is not a protocol.

**Pull request merged**